### ISSUE TITLE
connecting to secured socket

### ISSUE THREAD

**Issue by RajatGoyal** [ericchapman/ruby_wamp_client#3]
```
require 'wamp_client'

wsuri = 'wss://api2.poloniex.com:443/'
  options = {
      uri: wsuri,
      realm: 'realm1',
      headers: {Upgrade: 'websocket',
                Connection: 'Upgrade',}
  }
  connection = WampClient::Connection.new(options)

  connection.on_connect do
    puts 'connection established'
  end

  connection.on_challenge do |authmethod, extra|
    puts 'challenged'
  end

  connection.on_disconnect do |reason|
    puts 'disconnected'
  end

  connection.on_leave do |reason, details|
    puts 'session terminted'
  end
  
  connection.on_join do |session, details|
    puts 'session created'

    def got_ticker(args, kwargs, details)
      puts args
      puts kwargs
      puts details
    end

    session.subscribe('ticker', method(:got_ticker))
  end

connection.open
```

I am trying to connect to a secured socket using this library, it gets connected, but the session never gets made(on_join is never called), the api seems to be working fine with autobahn javascript.

**Comment by ericchapman** [comment]
Thanks Rajat.  I hadn't gotten a chance to test secure sockets yet so no surprise there.  Give me a day or so and I will resolve.

Please keep that router up so I can test against it.  I don't have a wss router up currently.

**Comment by ericchapman** [comment]
@RajatGoyal This actually isn't an SSL problem.  It doesn't appear your router/server is talking WAMP.  I turned on logging (see my version 0.0.4 release that I just put up.  Add "verbose: true" to the connection options) and I am getting the following

```
TX: HELLO > [1, "realm1", {:roles=>{:caller=>{:features=>{:caller_identification=>true, :progressive_call_results=>true}}, :callee=>{:features=>{:caller_identification=>true, :pattern_based_registration=>true, :shared_registration=>true, :progressive_call_results=>true, :registration_revocation=>true}}, :publisher=>{:features=>{:publisher_identification=>true, :subscriber_blackwhite_listing=>true, :publisher_exclusion=>true}}, :subscriber=>{:features=>{:publisher_identification=>true, :pattern_based_subscription=>true, :subscription_revocation=>true}}}, :agent=>"Ruby-WampClient-0.0.4", :authid=>nil, :authmethods=>nil}]
RX(non-wamp): {:error=>"Invalid command."}
RX(non-wamp): [1010]
RX(non-wamp): [1010]
RX(non-wamp): [1010]
RX(non-wamp): [1010]
RX(non-wamp): [1010]
...
```

It looks like your router/server doesn't understand "HELLO" which is why no session is being established.  It looks like you are however getting the "tick" that you intended.

What is your router/server?

**Comment by RajatGoyal** [comment]
@ericchapman Thanks for the quick reply, It's not my server, it's a cryptocurrency exchange and they have this router in their web socket, It is documented under push API on this link https://poloniex.com/support/api/, I tested the pastebin code (http://pastebin.com/dMX7mZE0) given their and it seems to be working.

**Comment by RajatGoyal** [comment]
@ericchapman It seems to be down currently, It was working using the pastebin code before, but it's not working now, and from the ruby code I posted originally, I was getting 'connection established' as output and nothing else.

**Comment by ericchapman** [comment]
@RajatGoyal Got ya.  I'll look a little more.  From first glance it isn't an SSL problem.  I am getting data back.  My guess is going to be that my library is missing a step in the negotiating and that is causing their system to NOT talk WAMP.  I'll look at the API documentation a bit.  I agree it should be talking WAMP from the JS Autobahn example.  I have only tested my code against the Crossbar.io router so it very well could be I am missing something in the handshake

**Comment by ericchapman** [comment]
@RajatGoyal I found it.  First off your uri and connection was wrong.  Should be

``` ruby
wsuri = 'wss://api.poloniex.com'
options = {
    uri: wsuri,
    verbose: true,
    realm: 'realm1'
}
```

Once I figured that out, I was then getting disconnected from the router.  When I viewed the reason, it was because the "authid" and "authmethods" were nil.  They actually shouldn't even have been included in the payload if they are nil which was an error on my part.  I released v0.0.5 fixing this.  Now I get the following when I include 'verbose'

```
connection established
TX: HELLO > [1, "realm1", {:roles=>{:caller=>{:features=>{:caller_identification=>true, :progressive_call_results=>true}}, :callee=>{:features=>{:caller_identification=>true, :pattern_based_registration=>true, :shared_registration=>true, :progressive_call_results=>true, :registration_revocation=>true}}, :publisher=>{:features=>{:publisher_identification=>true, :subscriber_blackwhite_listing=>true, :publisher_exclusion=>true}}, :subscriber=>{:features=>{:publisher_identification=>true, :pattern_based_subscription=>true, :subscription_revocation=>true}}}, :agent=>"Ruby-WampClient-0.0.5"}]
RX: WELCOME > [2, 8952133679609310, {:x_cb_node_id=>"frontend6.poloniex.com", :realm=>"realm1", :authid=>"3CVJ-43CQ-PNAL-XGMS-QPTR-GKUP", :authrole=>"public", :authmethod=>"anonymous", :authprovider=>"static", :roles=>{:broker=>{:features=>{:payload_encryption_cryptobox=>true, :payload_transparency=>true, :subscription_revocation=>true, :publisher_exclusion=>true, :subscriber_blackwhite_listing=>true, :subscription_meta_api=>true, :session_meta_api=>true, :pattern_based_subscription=>true, :publisher_identification=>true}}, :dealer=>{:features=>{:payload_encryption_cryptobox=>true, :payload_transparency=>true, :registration_revocation=>true, :progressive_call_results=>true, :shared_registration=>true, :registration_meta_api=>true, :session_meta_api=>true, :pattern_based_registration=>true, :caller_identification=>true}}}}]
session created
TX: SUBSCRIBE > [32, 4283005629906703, {}, "ticker"]
RX: SUBSCRIBED > [33, 4283005629906703, 4083579880050468]
RX: EVENT > [36, 4083579880050468, 454106336213194, {}, ["BTC_MAID", "0.00015181", "0.00015174", "0.00015021", "0.08520980", "661.98950416", "4451058.46862098", 0, "0.00015800", "0.00013625"]]
BTC_MAID
0.00015181
0.00015174
0.00015021
0.08520980
661.98950416
4451058.46862098
0
0.00015800
0.00013625
{}
{:publication=>454106336213194}
RX: EVENT > [36, 4083579880050468, 8476931001494907, {}, ["BTC_OMNI", "0.00841545", "0.00830848", "0.00806114", "-0.01041041", "89.41203564", "10068.08870915", 0, "0.00923996", "0.00803905"]]
BTC_OMNI
0.00841545
...
```

Update to v0.0.5 and the issue is resolved.  Thanks @RajatGoyal

**Comment by ericchapman** [comment]
@RajatGoyal Thank you and please do :)
